mudslide/models.py

```diff
@@ -553,12 +562,205 @@ def dV(self, X: ArrayLike) -> ArrayLike:
+            with open(self.coord_path, "r") as coord_file:
+                lines = coord_file.readlines()
+            read_start = np.argwhere(np.array(["$coord\n" in line for line in lines]))[0][0]
+
+        first_char = ""
+        current_line = read_start + 1
+
+        while True:
+            line = lines[current_line]
+            first_char = line[0]
+            if first_char == "$":
+                break
+            line_split = re.split("\s+", line.strip())
+            self.atom_order.append(line_split[-1])
+            coords.extend([float(val) for val in line_split[0:3]])
+            current_line += 1
+    
+        self.ndim_ = len(coords)
+        self.X = np.array(coords, dtype=np.float64).reshape(self.ndim())
+        self.mass = [masses[atom] for atom in self.atom_order]
+        self.mass = [3 * [masses[atom]] for atom in self.atom_order]
+        self.mass = np.array(self.mass, dtype=np.float64).reshape(self.ndim()) * amu_to_au
+
+
+    def update_coords(self, X):
+
+        first_space = 4 * " "
+        first_space_neg = 3 * " "
+
+        space = 6 * " "
+        space_neg = 5 * " "
+
+        X = X.reshape((self.ndim() // 3, 3))
+
+        with open(self.coord_path, "r") as coord_file:
+            lines = coord_file.readlines()
+
+
+            read_start = np.argwhere(np.array(["$coord\n" in line for line in lines]))[0][0]
+
+        with open(self.coord_path, "w") as coord_file:
+            current_line = read_start + 1
+
+            for i, coord in enumerate(X):  
+                if coord[0] < 0:
+                    s1 = first_space_neg
+                else:
+                    s1 = first_space
+
+                if coord[1] < 0:
+                    s2 = space_neg
+                else:
+                    s2 = space
+
+                if coord[2] < 0:
+                    s3 = space_neg
+                else:
+                    s3 = space
+                lines[current_line] = s1 + f"{coord[0]:1.14f}" + s2 + f"{coord[1]:1.14f}" + s3 + f"{coord[2]:1.14f}" + space + self.atom_order[i] + "\n"  
+                current_line+=1
+            coord_file.write("".join(lines))            
+
+
+    def call_and_parse_turbomole(self, outname="turbo.out"):
+        with open(outname, "w") as f:
+            ridft = subprocess.run("ridft", stdout=f)
+            rdgrad = subprocess.run("rdgrad", stdout=f)
+            egrad = subprocess.run("egrad", stdout=f)
+    
+        with open(outname, "r") as f:
+            data_dict = turboparse.parse_turbo(f)
+
+        parsed_nac_coupling = data_dict["egrad"]["coupling"]
+        
+        self.derivative_coupling = np.zeros((self.nstates(), self.nstates(), self.ndim()))
+        for i in range(self.nstates()):
+            for j in range(self.nstates()):
+                if i != j:
+                    for dct in parsed_nac_coupling:
+                        if (dct["bra_state"] == i) and (dct["ket_state"] == j):
```

Comment:
This is a wasteful and convoluted way to go through all the data. A better looping structure would just be
```
for dct in parsed_nac_coupling:
    i = dct["bra_state"]
    j = dct["ket_state"]
    ...
```

Then you've removed two loops and an unnecessary if branch.

Reply:
addressed!